lease/lessor.go

```diff
@@ -56,12 +58,39 @@ type Lessor interface {
 	// given lease will be removed. If the ID does not exist, an error
 	// will be returned.
 	Revoke(id LeaseID) error
+
+	// Promote promotes the lessor to be the primary lessor. Primary lessor manages
+	// the expiration and renew of leases.
+	Promote()
+
+	// Demote demotes the lessor from being the primary lessor.
+	Demote()
+
+	// Renew renews a lease with given ID.  If the ID does not exist, an error
+	// will be returned.
+	Renew(id LeaseID) error
 }
 
 // lessor implements Lessor interface.
 // TODO: use clockwork for testability.
 type lessor struct {
 	mu sync.Mutex
+
+	// primary indicates if this lessor is the primary lessor. The primary
+	// lessor manages lease expiration and renew.
+	//
+	// in etcd, raft leader is the primary. Thus There might be two primary
+	// leaders at the same time (raft allows concurrent leader but with different term)
+	// for at most a leader election timeout.
+	// The old primary leader cannot affect the correctness since its proposal has a
+	// smaller term and will not be committed.
+	//
+	// TODO: raft follower do not forward lease management proposals. There might be a
+	// very small window (within second normally which depends on go scheduling) that
+	// a raft follow is the primary between the raft leader demotion and lessor demotion.
+	// Usually this should not be a problem. Lease should not be that sensitive to timing.
+	primary bool
```

Comment:
What I'm getting at is whether `lessor` should know about Promote/Demote, having this split mode where it knows whether it's leader or follower and has to account for interface misuse based on a flag.

Instead of having promote/demote in the lessor interface, something like this could keep it simpler:
```go
type RaftLessor struct {
Lessor
primary Lessor
follower Lessor
}
NewRaftLessor(primary Lessor, follower Lessor) { return &RaftLessor{follower, primary, follower} }
(rl *RaftLessor) Promote() { rl.Lessor = primary }
(rl *RaftLessor) Demote() { rl.Lessor = follower}
```
where primary is the full-fledged version and follower is a more limited implementation